internal/servers/controller/listeners.go

```diff
@@ -168,6 +83,104 @@ func (c *Controller) startListeners(ctx context.Context) error {
 	return nil
 }
 
+func (c *Controller) configureForAPI(ctx context.Context, ln *base.ServerListener) ([]func(), error) {
+	apiServers := make([]func(), 0)
+
+	handler, err := c.apiHandler(HandlerProperties{
+		ListenerConfig: ln.Config,
+		CancelCtx:      ctx,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	cancelCtx := c.baseContext // Resolve to avoid race conditions if the base context is replaced.
+	server := &http.Server{
+		Handler:           handler,
+		ReadHeaderTimeout: 10 * time.Second,
+		ReadTimeout:       30 * time.Second,
+		IdleTimeout:       5 * time.Minute,
+		ErrorLog:          c.logger.StandardLogger(nil),
+		BaseContext:       func(net.Listener) context.Context { return cancelCtx },
+	}
+	ln.HTTPServer = server
+
+	if ln.Config.HTTPReadHeaderTimeout > 0 {
+		server.ReadHeaderTimeout = ln.Config.HTTPReadHeaderTimeout
+	}
+	if ln.Config.HTTPReadTimeout > 0 {
+		server.ReadTimeout = ln.Config.HTTPReadTimeout
+	}
+	if ln.Config.HTTPWriteTimeout > 0 {
+		server.WriteTimeout = ln.Config.HTTPWriteTimeout
+	}
+	if ln.Config.HTTPIdleTimeout > 0 {
+		server.IdleTimeout = ln.Config.HTTPIdleTimeout
+	}
+
+	switch ln.Config.TLSDisable {
+	case true:
+		l, err := ln.Mux.RegisterProto(alpnmux.NoProto, nil)
+		if err != nil {
+			return nil, fmt.Errorf("error getting non-tls listener: %w", err)
+		}
+		if l == nil {
+			return nil, errors.New("could not get non-tls listener")
+		}
+		apiServers = append(apiServers, func() { go server.Serve(l) })
+
+	default:
+		protos := []string{"", "http/1.1", "h2"}
+		for _, v := range protos {
+			l := ln.Mux.GetListener(v)
+			if l == nil {
+				return nil, fmt.Errorf("could not get tls proto %q listener", v)
+			}
+			apiServers = append(apiServers, func() { go server.Serve(l) })
```

Comment:
This is the problematic bit: We have multiple listeners - Therefore I wasn't able to use the same pattern here as I did on [`configureForCluster`](https://github.com/hashicorp/boundary/pull/1874/files#diff-954e2ccd0cd7c7d7991745a59b951b9149a750bdf8d18ee04e88be3e2524310fR70) (not returning anything and mutating the listener object and then using the new objects within it to start the server).

Possible solutions:
1. We change `base.ServerListener`'s `ALPNListener` field to be a slice of `net.Listener` instead of just one.
2. We add a new field to `base.ServerListener` which would be a slice of funcs to be used to startup that particular listener.

Thoughts very much welcome!

Reply:
https://github.com/hashicorp/boundary/pull/1874#discussion_r819607622